src/asci.cc

```diff
@@ -0,0 +1,1018 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Forte: an open-source plugin to Psi4 (https://github.com/psi4/psi4)
+ * that implements a variety of quantum chemistry methods for strongly
+ * correlated electrons.
+ *
+ * Copyright (c) 2012-2017 by its authors (see COPYING, COPYING.LESSER,
+ * AUTHORS).
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ *
+ * @END LICENSE
+ */
+
+#include "psi4/libpsi4util/process.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/pointgrp.h"
+#include "psi4/libpsio/psio.hpp"
+
+#include "../helpers/printing.h"
+#include "asci.h"
+
+using namespace psi;
+
+namespace psi {
+namespace forte {
+
+//#ifdef _OPENMP
+//#include <omp.h>
+//#else
+//#define omp_get_max_threads() 1
+//#define omp_get_thread_num() 0
+//#define omp_get_num_threads() 1
+//#endif
+
+void set_ASCI_options(ForteOptions& foptions) {
+    /* Convergence Threshold -*/
+    foptions.add_double("ASCI_CONVERGENCE", 1e-5, "ASCI Convergence threshold");
+    foptions.add_int("ASCI_MAX_CYCLE", 20, "ASCI MAX Cycle");
+    foptions.add_int("ASCI_TDET", 2000, "ASCI Max det");
+    foptions.add_int("ASCI_CDET", 200, "ASCI Max reference det");
+    foptions.add_double("ASCI_PRESCREEN_THRESHOLD", 1e-12, "ASCI PRESCREEN THRESH");
+}
+
+bool pairCompDescend(const std::pair<double, Determinant> E1, const std::pair<double, Determinant> E2) {
+    return E1.first > E2.first;
+}
+
+ASCI::ASCI(SharedWavefunction ref_wfn, Options& options,
```

Comment:
I see what's happening here. We can check this in, but in the future I would refactor this class and combine with ACI and HBCI. Perhaps one solution is inheritance, where you have a base SelectedCI class.

Reply:
I think a base SelectedCI class would simplify a lot of the code, and it would be pretty easy to implement. I'll address all of the specific changes here and open a new PR to establish a new organization and perhaps add an optimized HBCI implementation to that as well.